File: src/utils/fontManager.js
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const os = require('os');
const { getSystemFonts } = require('./systemFonts');

/**
 * Handles font activation, deactivation, and status checking
 * across different operating systems
 */
class FontManager {
  constructor() {
    this.platform = os.platform();
    this.activeFonts = [];
    this.fontCachePath = path.join(
      process.env.APPDATA || 
      (process.platform === 'darwin' ? 
        path.join(os.homedir(), 'Library/Application Support') : 
        path.join(os.homedir(), '.config')),
      'FontManager',
      'active-fonts.json'
    );
    
    // Create directory if it doesn't exist
    const dir = path.dirname(this.fontCachePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    
    // Create file if it doesn't exist
    if (!fs.existsSync(this.fontCachePath)) {
      fs.writeFileSync(this.fontCachePath, JSON.stringify([]));
    }
    
    // Load active fonts from cache
    this.loadActiveFonts();
  }
  
  /**
   * Load active fonts from cache file
   */
  loadActiveFonts() {
    try {
      const data = fs.readFileSync(this.fontCachePath, 'utf8');
      this.activeFonts = JSON.parse(data);
    } catch (error) {
      console.error('Error loading active fonts:', error);
      this.activeFonts = [];
    }
  }
  
  /**
   * Save active fonts to cache file
   */
  saveActiveFonts() {
    try {
      fs.writeFileSync(this.fontCachePath, JSON.stringify(this.activeFonts));
    } catch (error) {
      console.error('Error saving active fonts:', error);
    }
  }
  
  /**
   * Activate a font
   * @param {string} fontPath - Path to the font file
   * @returns {Promise<boolean>} - Success status
   */
  async activateFont(fontPath) {
    if (!fontPath) {
      throw new Error('Font path is required');
    }
    
    if (!fs.existsSync(fontPath)) {
      throw new Error(`Font file does not exist: ${fontPath}`);
    }
    
    switch (this.platform) {
      case 'win32':
        return await this.activateFontWindows(fontPath);
      case 'darwin':
        return await this.activateFontMacOS(fontPath);
      case 'linux':
        return await this.activateFontLinux(fontPath);
      default:
        throw new Error(`Unsupported platform: ${this.platform}`);
    }
  }
  
  /**
   * Deactivate a font
   * @param {string} fontPath - Path to the font file
   * @returns {Promise<boolean>} - Success status
   */
  async deactivateFont(fontPath) {
    if (!fontPath) {
      throw new Error('Font path is required');
    }
    
    switch (this.platform) {
      case 'win32':
        return await this.deactivateFontWindows(fontPath);
      case 'darwin':
        return await this.deactivateFontMacOS(fontPath);
      case 'linux':
        return await this.deactivateFontLinux(fontPath);
      default:
        throw new Error(`Unsupported platform: ${this.platform}`);
    }
  }
  
  /**
   * Check if a font is active
   * @param {string} fontPath - Path to the font file
   * @returns {boolean} - True if active
   */
  isFontActive(fontPath) {
    return this.activeFonts.includes(fontPath);
  }
  
  /**
   * Get all active fonts
   * @returns {Array<string>} - Array of active font paths
   */
  async getActiveFonts() {
    const systemFonts = await getSystemFonts();
    const activeSystemFonts = systemFonts.filter(font => font.active);
    
    // Get font details from active font paths
    const activeFontDetails = this.activeFonts.map(fontPath => {
      const fileName = path.basename(fontPath);
      const fontNameMatch = fileName.match(/^(.+)\.(ttf|otf|woff|woff2)$/i);
      const fontName = fontNameMatch ? fontNameMatch[1] : fileName;
      
      return {
        family: fontName.replace(/[_-]/g, ' '),
        path: fontPath,
        active: true,
        type: 'local',
        id: fontPath
      };
    });
    
    return [...activeSystemFonts, ...activeFontDetails];
  }
  
  /**
   * Activate a font on Windows
   * @param {string} fontPath - Path to the font file
   * @returns {Promise<boolean>} - Success status
   */
  async activateFontWindows(fontPath) {
    // On Windows, we need to copy the font to the Windows font directory
    const fontFileName = path.basename(fontPath);
    const windowsFontDir = path.join(process.env.WINDIR, 'Fonts');
    const destFontPath = path.join(windowsFontDir, fontFileName);
    
    return new Promise((resolve, reject) => {
      // Check if font is already installed
      if (fs.existsSync(destFontPath)) {
        if (!this.activeFonts.includes(fontPath)) {
          this.activeFonts.push(fontPath);
          this.saveActiveFonts();
        }
        resolve(true);
        return;
      }
      
      // Copy font file to Windows font directory
      fs.copyFile(fontPath, destFontPath, (err) => {
        if (err) {
          reject(err);
          return;
        }
        
        // Add registry entry (requires admin rights)
        const regCommand = `reg add "HKLM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\Fonts" /v "${fontFileName}" /t REG_SZ /d "${destFontPath}" /f`;
        exec(regCommand, (error) => {
          if (error) {
            console.warn('Failed to add font to registry (may require admin rights):', error);
          }
          
          // Add to active fonts list
          if (!this.activeFonts.includes(fontPath)) {
            this.activeFonts.push(fontPath);
            this.saveActiveFonts();
          }
          
          resolve(true);
        });
      });
    });
  }
  
  /**
   * Deactivate a font on Windows
   * @param {string} fontPath - Path to the font file
   * @returns {Promise<boolean>} - Success status
   */
  async deactivateFontWindows(fontPath) {
    const fontFileName = path.basename(fontPath);
    const windowsFontDir = path.join(process.env.WINDIR, 'Fonts');
    const destFontPath = path.join(windowsFontDir, fontFileName);
    
    return new Promise((resolve, reject) => {
      // Check if font exists in Windows font directory
      if (!fs.existsSync(destFontPath)) {
        // Remove from active fonts list
        this.activeFonts = this.activeFonts.filter(f => f !== fontPath);
        this.saveActiveFonts();
        resolve(true);
        return;
      }
      
      // Remove registry entry (requires admin rights)
      const regCommand = `reg delete "HKLM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\Fonts" /v "${fontFileName}" /f`;
      exec(regCommand, (error) => {
        if (error) {
          console.warn('Failed to remove font from registry (may require admin rights):', error);
        }
        
        // Delete font file from Windows font directory
        fs.unlink(destFontPath, (err) => {
          if (err) {
            reject(err);
            return;
          }
          
          // Remove from active fonts list
          this.activeFonts = this.activeFonts.filter(f => f !== fontPath);
          this.saveActiveFonts();
          
          resolve(true);
        });
      });
    });
  }
  
  /**
   * Activate a font on macOS
   * @param {string} fontPath - Path to the font file
   * @returns {Promise<boolean>} - Success status
   */
  async activateFontMacOS(fontPath) {
    const userFontDir = path.join(os.homedir(), 'Library/Fonts');
    const fontFileName = path.basename(fontPath);
    const destFontPath = path.join(userFontDir, fontFileName);
    
    return new Promise((resolve, reject) => {
      // Check if font is already installed
      if (fs.existsSync(destFontPath)) {
        if (!this.activeFonts.includes(fontPath)) {
          this.activeFonts.push(fontPath);
          this.saveActiveFonts();
        }
        resolve(true);
        return;
      }
      
      // Copy font file to user font directory
      fs.copyFile(fontPath, destFontPath, (err) => {
        if (err) {
          reject(err);
          return;
        }
        
        // Add to active fonts list
        if (!this.activeFonts.includes(fontPath)) {
          this.activeFonts.push(fontPath);
          this.saveActiveFonts();
        }
        
        resolve(true);
      });
    });
  }
  
  /**
   * Deactivate a font on macOS
   * @param {string} fontPath - Path to the font file
   * @returns {Promise<boolean>} - Success status
   */
  async deactivateFontMacOS(fontPath) {
    const userFontDir = path.join(os.homedir(), 'Library/Fonts');
    const fontFileName = path.basename(fontPath);
    const destFontPath = path.join(userFontDir, fontFileName);
    
    return new Promise((resolve, reject) => {
      // Check if font exists in user font directory
      if (!fs.existsSync(destFontPath)) {
        // Remove from active fonts list
        this.activeFonts = this.activeFonts.filter(f => f !== fontPath);
        this.saveActiveFonts();
        resolve(true);
        return;
      }
      
      // Delete font file from user font directory
      fs.unlink(destFontPath, (err) => {
        if (err) {
          reject(err);
          return;
        }
        
        // Remove from active fonts list
        this.activeFonts = this.activeFonts.filter(f => f !== fontPath);
        this.saveActiveFonts();
        
        resolve(true);
      });
    });
  }
  
  /**
   * Activate a font on Linux
   * @param {string} fontPath - Path to the font file
   * @returns {Promise<boolean>} - Success status
   */
  async activateFontLinux(fontPath) {
    const userFontDir = path.join(os.homedir(), '.local/share/fonts');
    if (!fs.existsSync(userFontDir)) {
      fs.mkdirSync(userFontDir, { recursive: true });
    }
    
    const fontFileName = path.basename(fontPath);
    const destFontPath = path.join(userFontDir, fontFileName);
    
    return new Promise((resolve, reject) => {
      // Check if font is already installed
      if (fs.existsSync(destFontPath)) {
        if (!this.activeFonts.includes(fontPath)) {
          this.activeFonts.push(fontPath);
          this.saveActiveFonts();
        }
        resolve(true);
        return;
      }
      
      // Copy font file to user font directory
      fs.copyFile(fontPath, destFontPath, (err) => {
        if (err) {
          reject(err);
          return;
        }
        
        // Refresh font cache
        exec('fc-cache -f', (error) => {
          if (error) {
            console.warn('Failed to refresh font cache:', error);
          }
          
          // Add to active fonts list
          if (!this.activeFonts.includes(fontPath)) {
            this.activeFonts.push(fontPath);
            this.saveActiveFonts();
          }
          
          resolve(true);
        });
      });
    });
  }
  
  /**
   * Deactivate a font on Linux
   * @param {string} fontPath - Path to the font file
   * @returns {Promise<boolean>} - Success status
   */
  async deactivateFontLinux(fontPath) {
    const userFontDir = path.join(os.homedir(), '.local/share/fonts');
    const fontFileName = path.basename(fontPath);
    const destFontPath = path.join(userFontDir, fontFileName);
    
    return new Promise((resolve, reject) => {
      // Check if font exists in user font directory
      if (!fs.existsSync(destFontPath)) {
        // Remove from active fonts list
        this.activeFonts = this.activeFonts.filter(f => f !== fontPath);
        this.saveActiveFonts();
        resolve(true);
        return;
      }
      
      // Delete font file from user font directory
      fs.unlink(destFontPath, (err) => {
        if (err) {
          reject(err);
          return;
        }
        
        // Refresh font cache
        exec('fc-cache -f', (error) => {
          if (error) {
            console.warn('Failed to refresh font cache:', error);
          }
          
          // Remove from active fonts list
          this.activeFonts = this.activeFonts.filter(f => f !== fontPath);
          this.saveActiveFonts();
          
          resolve(true);
        });
      });
    });
  }
}

module.exports = new FontManager();

File: python_helpers/font_manager.py
#!/usr/bin/env python3
"""
Font Manager Script - Python helper for font installation/uninstallation
This script provides cross-platform functionality to install and uninstall fonts
"""

import os
import sys
import json
import shutil
import platform
import subprocess
from pathlib import Path

def get_user_fonts_directory():
    """Get the user fonts directory based on the operating system"""
    system = platform.system()
    
    if system == 'Windows':
        return os.path.join(os.environ.get('WINDIR', 'C:\\Windows'), 'Fonts')
    elif system == 'Darwin':  # macOS
        return os.path.join(os.path.expanduser('~'), 'Library/Fonts')
    elif system == 'Linux':
        return os.path.join(os.path.expanduser('~'), '.local/share/fonts')
    else:
        return os.path.join(os.path.expanduser('~'), 'fonts')

def activate_font(font_path):
    """
    Installs a font on the system
    Args:
        font_path: Path to the font file
    Returns:
        bool: Success status
    """
    system = platform.system()
    
    try:
        if system == 'Windows':
            return activate_font_windows(font_path)
        elif system == 'Darwin':  # macOS
            return activate_font_macos(font_path)
        elif system == 'Linux':
            return activate_font_linux(font_path)
        else:
            print(f"Unsupported operating system: {system}")
            return False
    except Exception as e:
        print(f"Error activating font: {e}")
        return False

def deactivate_font(font_path):
    """
    Removes a font from the system
    Args:
        font_path: Path to the font file
    Returns:
        bool: Success status
    """
    system = platform.system()
    
    try:
        if system == 'Windows':
            return deactivate_font_windows(font_path)
        elif system == 'Darwin':  # macOS
            return deactivate_font_macos(font_path)
        elif system == 'Linux':
            return deactivate_font_linux(font_path)
        else:
            print(f"Unsupported operating system: {system}")
            return False
    except Exception as e:
        print(f"Error deactivating font: {e}")
        return False

def activate_font_windows(font_path):
    """
    Install a font on Windows
    Uses the system font API through PowerShell
    """
    try:
        # Create a PowerShell command to add the font
        font_file = os.path.basename(font_path)
        user_fonts_dir = get_user_fonts_directory()
        
        # Copy the font to the Fonts directory
        dest_path = os.path.join(user_fonts_dir, font_file)
        
        if not os.path.exists(dest_path):
            shutil.copy2(font_path, dest_path)
        
        # Use PowerShell to install the font through the Shell API
        ps_command = f"""
        $objShell = New-Object -ComObject Shell.Application
        $objFolder = $objShell.Namespace(0x14)
        $objFolder.CopyHere("{dest_path}")
        """
        
        # Execute PowerShell command
        subprocess.run(['powershell', '-Command', ps_command], 
                      check=True, 
                      capture_output=True)
        
        return True
    except Exception as e:
        print(f"Error activating font on Windows: {e}")
        return False

def deactivate_font_windows(font_path):
    """
    Uninstall a font on Windows
    Removes from user fonts directory and updates registry
    """
    try:
        font_file = os.path.basename(font_path)
        user_fonts_dir = get_user_fonts_directory()
        
        # Path to the font in the Fonts directory
        dest_path = os.path.join(user_fonts_dir, font_file)
        
        # Check if the font exists in the Fonts directory
        if os.path.exists(dest_path):
            # Use PowerShell to properly uninstall the font
            ps_command = f"""
            Remove-ItemProperty -Path "HKCU:\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\Fonts" -Name "{font_file}"
            Remove-Item -Path "{dest_path}" -Force
            """
            
            # Execute PowerShell command
            subprocess.run(['powershell', '-Command', ps_command], 
                          check=True, 
                          capture_output=True)
        
        return True
    except Exception as e:
        print(f"Error deactivating font on Windows: {e}")
        return False

def activate_font_macos(font_path):
    """
    Install a font on macOS
    Copies to user font directory and updates font cache
    """
    try:
        font_file = os.path.basename(font_path)
        user_fonts_dir = get_user_fonts_directory()
        
        # Ensure user fonts directory exists
        os.makedirs(user_fonts_dir, exist_ok=True)
        
        # Copy the font to the user fonts directory
        dest_path = os.path.join(user_fonts_dir, font_file)
        
        if not os.path.exists(dest_path):
            shutil.copy2(font_path, dest_path)
        
        # Update font cache
        subprocess.run(['atsutil', 'databases', '-remove'], 
                      check=True, 
                      capture_output=True)
        
        return True
    except Exception as e:
        print(f"Error activating font on macOS: {e}")
        return False

def deactivate_font_macos(font_path):
    """
    Uninstall a font on macOS
    Removes from user font directory and updates font cache
    """
    try:
        font_file = os.path.basename(font_path)
        user_fonts_dir = get_user_fonts_directory()
        
        # Path to the font in the user fonts directory
        dest_path = os.path.join(user_fonts_dir, font_file)
        
        # Check if the font exists in the user fonts directory
        if os.path.exists(dest_path):
            os.remove(dest_path)
        
        # Update font cache
        subprocess.run(['atsutil', 'databases', '-remove'], 
                      check=True, 
                      capture_output=True)
        
        return True
    except Exception as e:
        print(f"Error deactivating font on macOS: {e}")
        return False

def activate_font_linux(font_path):
    """
    Install a font on Linux
    Copies to user font directory and updates font cache
    """
    try:
        font_file = os.path.basename(font_path)
        user_fonts_dir = get_user_fonts_directory()
        
        # Ensure user fonts directory exists
        os.makedirs(user_fonts_dir, exist_ok=True)
        
        # Copy the font to the user fonts directory
        dest_path = os.path.join(user_fonts_dir, font_file)
        
        if not os.path.exists(dest_path):
            shutil.copy2(font_path, dest_path)
        
        # Update font cache
        subprocess.run(['fc-cache', '-f'], 
                      check=True, 
                      capture_output=True)
        
        return True
    except Exception as e:
        print(f"Error activating font on Linux: {e}")
        return False

def deactivate_font_linux(font_path):
    """
    Uninstall a font on Linux
    Removes from user font directory and updates font cache
    """
    try:
        font_file = os.path.basename(font_path)
        user_fonts_dir = get_user_fonts_directory()
        
        # Path to the font in the user fonts directory
        dest_path = os.path.join(user_fonts_dir, font_file)
        
        # Check if the font exists in the user fonts directory
        if os.path.exists(dest_path):
            os.remove(dest_path)
        
        # Update font cache
        subprocess.run(['fc-cache', '-f'], 
                      check=True, 
                      capture_output=True)
        
        return True
    except Exception as e:
        print(f"Error deactivating font on Linux: {e}")
        return False

def list_installed_fonts():
    """
    List all fonts installed on the system
    Returns:
        list: List of font paths
    """
    user_fonts_dir = get_user_fonts_directory()
    fonts = []
    
    try:
        # For the web app demo, simplify by listing fonts in the user fonts directory
        if os.path.exists(user_fonts_dir):
            for file in os.listdir(user_fonts_dir):
                file_path = os.path.join(user_fonts_dir, file)
                if os.path.isfile(file_path) and file.lower().endswith(('.ttf', '.otf', '.woff', '.woff2')):
                    fonts.append(file_path)
        
        # For the web app demo, add some sample fonts if none are found
        if not fonts:
            system = platform.system()
            if system == 'Windows':
                fonts = [
                    os.path.join(user_fonts_dir, 'Arial.ttf'),
                    os.path.join(user_fonts_dir, 'Times New Roman.ttf'),
                    os.path.join(user_fonts_dir, 'Calibri.ttf')
                ]
            elif system == 'Darwin':  # macOS
                fonts = [
                    os.path.join(user_fonts_dir, 'Arial.ttf'),
                    os.path.join(user_fonts_dir, 'Times New Roman.ttf'),
                    os.path.join(user_fonts_dir, 'Helvetica.ttf')
                ]
            elif system == 'Linux':
                fonts = [
                    os.path.join(user_fonts_dir, 'DejaVuSans.ttf'),
                    os.path.join(user_fonts_dir, 'FreeSans.ttf'),
                    os.path.join(user_fonts_dir, 'Liberation Sans.ttf')
                ]
    except Exception as e:
        print(f"Error listing installed fonts: {e}")
    
    return fonts

def main():
    """Main CLI function"""
    if len(sys.argv) < 2:
        print(json.dumps({
            'error': 'Missing command',
            'usage': 'python font_manager.py [activate|deactivate|list] [font_path]'
        }))
        return
    
    command = sys.argv[1]
    
    try:
        if command == 'activate' and len(sys.argv) >= 3:
            font_path = sys.argv[2]
            success = activate_font(font_path)
            print(json.dumps({'success': success}))
        
        elif command == 'deactivate' and len(sys.argv) >= 3:
            font_path = sys.argv[2]
            success = deactivate_font(font_path)
            print(json.dumps({'success': success}))
        
        elif command == 'list':
            fonts = list_installed_fonts()
            print(json.dumps({'fonts': fonts}))
        
        else:
            print(json.dumps({
                'error': f'Unknown command: {command}',
                'usage': 'python font_manager.py [activate|deactivate|list] [font_path]'
            }))
    
    except Exception as e:
        print(json.dumps({'error': str(e)}))

if __name__ == '__main__':
    main()